src/providers/mod.rs

```diff
@@ -0,0 +1,76 @@
+use failure::Error;
+use std::collections::HashMap;
+
+pub struct LaunchDescriptor<M: MachineSetup> {
+    pub region: M::Region,
+    pub log: slog::Logger,
+    pub max_wait: Option<std::time::Duration>,
+    pub machines: Vec<(String, M)>,
+}
+
+/// This is used to group machines into connections
+/// to cloud providers. e.g., for AWS we need a separate
+/// connection to each region.
+pub trait MachineSetup: Send + Clone {
```

Comment:
I think I'd prefer these on `spawn` rather than directly on the trait. Makes mocking a little easier.